providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1551 @@
+        G(v2, v7, v8, v13);                                                    \
+        G(v3, v4, v9, v14);                                                    \
+    } while ((void)0, 0)
+
+# undef PERMUTATION_P_COLUMN
+# define PERMUTATION_P_COLUMN(x, i)                                            \
+    do {                                                                       \
+        uint64_t *base = &x[16 * i];                                           \
+        PERMUTATION_P(                                                         \
+            *base,        *(base + 1),  *(base + 2),  *(base + 3),             \
+            *(base + 4),  *(base + 5),  *(base + 6),  *(base + 7),             \
+            *(base + 8),  *(base + 9),  *(base + 10), *(base + 11),            \
+            *(base + 12), *(base + 13), *(base + 14), *(base + 15)             \
+        );                                                                     \
+    } while ((void)0, 0)
+
+# undef PERMUTATION_P_ROW
+# define PERMUTATION_P_ROW(x, i)                                               \
+    do {                                                                       \
+        uint64_t *base = &x[2 * i];                                            \
+        PERMUTATION_P(                                                         \
+            *base,        *(base + 1),  *(base + 16),  *(base + 17),           \
+            *(base + 32), *(base + 33), *(base + 48),  *(base + 49),           \
+            *(base + 64), *(base + 65), *(base + 80),  *(base + 81),           \
+            *(base + 96), *(base + 97), *(base + 112), *(base + 113)           \
+        );                                                                     \
+    } while ((void)0, 0)
+
+typedef struct {
+    uint64_t v[ARGON2_QWORDS_IN_BLOCK];
+} BLOCK;
+
+typedef enum {
+    ARGON2_VERSION_10 = 0x10,
+    ARGON2_VERSION_13 = 0x13,
+    ARGON2_VERSION_NUMBER = ARGON2_VERSION_13
+} ARGON2_VERSION;
+
+typedef enum {
+    ARGON2_D  = 0,
+    ARGON2_I  = 1,
+    ARGON2_ID = 2
+} ARGON2_TYPE;
+
+typedef struct {
+    uint8_t *out;
+    uint32_t outlen;
+    uint8_t *pwd;
+    uint32_t pwdlen;
+    uint8_t *salt;
+    uint32_t saltlen;
+    uint8_t *secret;
+    uint32_t secretlen;
+    uint8_t *ad;
+    uint32_t adlen;
+    uint32_t t_cost;
+    uint32_t m_cost;
+    uint32_t lanes;
+    uint32_t threads;
+    uint32_t version;
+    uint32_t flags;
+    ARGON2_TYPE type;
+    BLOCK *memory;
+    uint32_t passes;
+    uint32_t memory_blocks;
+    uint32_t segment_length;
+    uint32_t lane_length;
+    OSSL_LIB_CTX *libctx;
+    EVP_MD *md;
+    EVP_MAC *mac;
+} ARGON2_CTX;
+
+typedef struct {
+    uint32_t pass;
+    uint32_t lane;
+    uint8_t slice;
+    uint32_t index;
+} ARGON2_POS;
+
+typedef struct {
+    ARGON2_POS pos;
+    ARGON2_CTX *ctx;
+} ARGON2_THREAD_DATA;
+
+union endianity_test {
+    long one;
+    char little;
+};
+
+static union endianity_test is_endian = { 1 };
```

Comment:
OpenSSL has an endian test, but to use them I think.

See include/internal/endian.h
`DECLARE_IS_ENDIAN;` declares it
`IS_LITTLE_ENDIAN` and `IS_BIG_ENDIAN` test it.